### PR TITLE
interfaces: use udev spec

### PR DESCRIPTION
**Pull request by stolowski**
Switch udev backend to the new spec-based API, updated affected interfaces and their test.

### CODE REVIEW

File: interfaces/udev/backend.go [hunk 1]
```diff
@@ -129,58 +126,14 @@ func (b *Backend) Remove(snapName string) error {
...
-			snapSnippets[string(snippet)] = snippet
-		}
-	}
-
-	var combinedSnippets [][]byte
-	for _, snippet := range snapSnippets {
-		combinedSnippets = append(combinedSnippets, snippet)
+func (b *Backend) deriveContent(spec *Specification, snapInfo *snap.Info) (content []string) {
+	for _, snippet := range spec.Snippets() {
+		content = append(content, snippet)
```
Comment (niemeyer) [inline interfaces/udev/backend.go]: Just in case: do we have proper newlines between the snippets here?

File: interfaces/udev/spec.go [hunk 2]
```diff
@@ -30,20 +32,19 @@ type Specification struct {
...
 }
 
 // Snippets returns a copy of all the snippets added so far.
-func (spec *Specification) Snippets() map[string]bool {
-	result := make(map[string]bool, len(spec.snippets))
-	for k, v := range spec.snippets {
-		result[k] = v
+func (spec *Specification) Snippets() (result []string) {
+	for k := range spec.snippets {
+		result = append(result, k)
```
Comment (niemeyer) [inline interfaces/udev/spec.go]: Where are we adding the newlines?  Not here, not in the backend.. do we miss it?

What about other backends?

Comment (stolowski) [inline interfaces/udev/spec.go]: All other backends add newlines either in the backend's deriveContent or SnippetForTag() method. So yes, it's a little bit inconsistent in this respect.

File: interfaces/udev/backend.go (hunk 1 above)
Comment (stolowski) [inline interfaces/udev/backend.go]: We're adding newlines in udev/backend.go in the "for _, snippet := range content {.." loop, which iterates over the list returned by deriveContent method.

File: interfaces/builtin/framebuffer.go [hunk 3]
```diff
@@ -81,19 +81,17 @@ func (iface *FramebufferInterface) AppArmorConnectedPlug(spec *apparmor.Specific
 	return nil
 }
 
+func (iface *FramebufferInterface) UdevConnectedPlug(spec *udev.Specification, plug *interfaces.Plug, slot *interfaces.Slot) error {
+	const udevRule = `KERNEL=="fb[0-9]*", TAG+="%s"`
+	for appName := range plug.Apps {
+		tag := udevSnapSecurityName(plug.Snap.Name(), appName)
+		spec.AddSnippet(fmt.Sprintf(udevRule, tag))
```
Comment (stolowski) [inline interfaces/builtin/framebuffer.go]: Right. I agree, good idea.

File: interfaces/builtin/hidraw.go [hunk 4]
```diff
@@ -113,24 +113,25 @@ func (iface *HidrawInterface) SanitizePlug(plug *interfaces.Plug) error {
...
+	}
+	usbProduct, pOk := slot.Attrs["usb-product"].(int64)
+	if !pOk {
+		return nil
+	}
+	path, ok := slot.Attrs["path"].(string)
+	if !ok || path == "" {
+		return nil
+	}
+	spec.AddSnippet(string(udevUsbDeviceSnippet("hidraw", usbVendor, usbProduct, "SYMLINK", strings.TrimPrefix(path, "/dev/"))))
```
Comment (stolowski) [inline interfaces/builtin/hidraw.go]: Ah, yes, in fact I changed udevUsbDeviceSnippet to return string at the end, just forgot to revisit this line. Thanks, fixed now.

**Pull request merged**